File: src/ios-objc/TwilioVideoPlugin.m
/********* TwilioVideo.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#import "TwilioVideoViewController.h"

@interface XXTwilioVideoPlugin : CDVPlugin

@end

@implementation XXTwilioVideoPlugin

- (void)openRoom:(CDVInvokedUrlCommand*)command {
    NSString* token = command.arguments[0];
    NSString* room = command.arguments[1];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIStoryboard *sb = [UIStoryboard storyboardWithName:@"TwilioVideo" bundle:nil];
        XXTwilioVideoViewController *vc = [sb instantiateViewControllerWithIdentifier:@"XXTwilioVideoViewController"];

        // UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:vc];
        // [vc.navigationItem setRightBarButtonItem:[[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(dismissTwilioVideoController)]];

        [self.viewController presentViewController:vc animated:YES completion:^{
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"ok"];
            [vc connectToRoom:room withToken:token];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    });

}

- (void) dismissTwilioVideoController {
    [self.viewController dismissViewControllerAnimated:YES completion:nil];
}

@end

File: src/ios/TwilioVideoPlugin.swift
/********* TwilioVideoPlugin.swift Cordova Plugin Implementation *******/

import Foundation
import TwilioVideo
import UIKit

@objc(TwilioVideoPlugin)
class TwilioVideoPlugin: CDVPlugin {
    var vc: TwilioVideoViewController?
    var sb: UIStoryboard?
    func openRoom(_ command: CDVInvokedUrlCommand) {
        let token: String = command.argument(at: 0, withDefault: "") as! String
        let room: String = command.argument(at: 1, withDefault: "") as! String
        self.commandDelegate.run(inBackground: {
            DispatchQueue.main.async(execute: {() -> Void in
                self.sb = UIStoryboard(name: "TwilioVideoSwift", bundle: nil)
                self.vc = self.sb!.instantiateViewController(withIdentifier: "TwilioVideoViewController") as? TwilioVideoViewController
                self.vc?.delegate = self
                self.viewController.present(self.vc!, animated: true, completion: {() -> Void in
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "ok")
                    self.vc!.connect(roomName: room, accessToken: token)
                    self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                })
            })
        })
    }
    
    override func onReset() {
        
    }
    
    func closeRoom(_ command: CDVInvokedUrlCommand) {
        if vc != nil  {
            vc!.disconnect()
        }
        dismissTwilioVideoController()
    }
    
    
    func dismissTwilioVideoController() {
       viewController.dismiss(animated: true) {() -> Void in }
    }
}

extension TwilioVideoPlugin : TwilioVideoViewControllerDelegate {
    func closeAfterDisconnect() {
        self.dismissTwilioVideoController()
    }
}
